### ISSUE TITLE
More granular hierarchy

### ISSUE THREAD

**Issue by joneshf** [FormationAI/dhall-bhat#56]
How would we feel about making the `Functor` hierarchy a little more granular? I.e. following PS's lead rather than Haskell's. We could go from:

![old](https://user-images.githubusercontent.com/1356417/52182476-6db8cb80-27b2-11e9-88f0-fae8f2f2e6f5.png)

to:

![new](https://user-images.githubusercontent.com/1356417/52182475-6db8cb80-27b2-11e9-9d47-d3963a88e833.png)

The impetus for this issue is that when trying to address #52, it's worlds easier if we have `List/extend`/`List/duplicate`. `List` cannot be a `Comonad` (because there's no way to implement `extract`). `List` can be a very useful  `Extend` though. There are many other data types that make great `Apply`s/`Bind`s/`Extend`s that don't necessarily have the right structure to implement `Applicative`/`Monad`/`Comonad`.

**Comment by sellout** [comment]
Yeah, I am all for this. `map2` (aka `liftA2`), `join`, and `duplicate` are the semigroup operation of the `Applicative`, `Monad`, and `Comonad` (co)monoids, respectively. And `ap`, `bind`, and `extend` can be defined in terms of those semigroup operations, so it’s a reasonable separation.

(Ideally, I’d say the type classes should have `map2`, `join`, and `deplicate` as their operations, with the others derived, but reasonable people differ.)

**Comment by FintanH** [comment]
Ya I'm totally up for this for 3 reasons
1. It will make defining dhall-bhat in terms of kind-poly easier
2. Can document as we split out these main pieces
3. You're dead right about functionality around `extend` and `duplicate`

**Comment by FintanH** [comment]
Is worth having `Pointed` as well? The graph above would be `Applicative <- (Pointed, Apply)`

laws:
* `fmap f . point === point . f`
* `fmap point === point`

**Comment by joneshf** [comment]
I'll have a think about the first one, but the second one seems like it might not always hold. E.g., if `point :: a -> [a]` then:

```Haskell
fmap point [1, 2, 3]
  = [[1], [2], [3]]
  /= [[1, 2, 3]]
  = point [1, 2, 3]
```

**Comment by FintanH** [comment]
Ah good _point_! I was thinking of `Optional` when I was riffing on it.

**Comment by FintanH** [comment]
lmao https://wiki.haskell.org/Why_not_Pointed%3F

> (he still provides the pointed package only because “people were whinging”).

**Comment by FintanH** [comment]
Seems like it's not as useful as I thought according to Kmett, so maybe it's best left out unless you can think of anything :) fwiw Kmett says the first law is a free theorem so that's all G :ok_hand:

**Comment by sellout** [comment]
Yeah, I’m sympathetic to wanting `Pointed`, but it ends up being “`Default` in the category of endofunctors”.

**Comment by FintanH** [comment]
Ran into a roadblock when prototyping this work.

We can't do `Applicative f //\\ Bind f` because they share fields. We also can't project fields for type records :( 

One workaround is have `Monad/Type` be defined as `Bind f //\\ { pure : forall(a : Type) -> a -> f a }`

**Comment by joneshf** [comment]
Huh, I wonder why there isn't a "prefer" operator for types–like `//` for terms.
